test/integration/cds_integration_test.cc

```diff
@@ -222,6 +222,123 @@ TEST_P(CdsIntegrationTest, CdsClusterTeardownWhileConnecting) {
   EXPECT_LE(cx_counter->value(), 1);
 }
 
+class LazyInitClusterStatsTest : public CdsIntegrationTest {
+
+public:
+  void initialize() {
+    this->use_real_stats_ = true;
+    config_helper_.addConfigModifier([](::envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+      bootstrap.set_enable_lazyinit_stats(true);
+    });
+    CdsIntegrationTest::initialize();
+    test_server_->waitForCounterGe("cluster_manager.cluster_added", 1);
+  }
+
+  void SendRequestToCluster1AndWaitForResponse() {
+    BufferingStreamDecoderPtr response = IntegrationUtil::makeSingleRequest(
+        lookupPort("http"), "GET", "/cluster1", "", downstream_protocol_, version_, "foo.com");
+    ASSERT_TRUE(response->complete());
+    cleanupUpstreamAndDownstream();
+  };
+
+  void ExpectStats(absl::string_view cluster, bool initialized_gauge_exists, int initialized_gauge,
+                   bool cx_active_exists, int upstream_cx_active) {
+    if (initialized_gauge_exists) {
+      EXPECT_EQ(
+          test_server_->gauge(absl::StrCat("cluster.", cluster, ".ClusterTrafficStats.initialized"))
+              ->value(),
+          initialized_gauge);
+    } else {
+      EXPECT_EQ(test_server_->gauge(
+                    absl::StrCat("cluster.", cluster, ".ClusterTrafficStats.initialized")),
+                nullptr);
+    }
+    if (cx_active_exists) {
+      EXPECT_EQ(
+          test_server_->counter(absl::StrCat("cluster.", cluster, ".upstream_cx_total"))->value(),
+          upstream_cx_active);
+    } else {
+      EXPECT_EQ(test_server_->counter(absl::StrCat("cluster.", cluster, ".upstream_cx_total")),
+                nullptr);
+    }
+  }
+  void UpdateCluster1() {
+    envoy::config::cluster::v3::Cluster cluster1_updated = cluster_creator_(
+        ClusterName1, fake_upstreams_[UpstreamIndex2]->localAddress()->ip()->port(),
+        Network::Test::getLoopbackAddressString(ipVersion()), "ROUND_ROBIN");
+    sendDiscoveryResponse<envoy::config::cluster::v3::Cluster>(
+        Config::TypeUrl::get().Cluster, {cluster1_updated}, {cluster1_updated}, {}, "42");
+  }
+};
+INSTANTIATE_TEST_SUITE_P(IpVersionsClientTypeDelta, LazyInitClusterStatsTest,
+                         DELTA_SOTW_GRPC_CLIENT_INTEGRATION_PARAMS);
+
+// Test that LazyInitTrafficStats gets created and updated correctly.
+TEST_P(LazyInitClusterStatsTest, LazyInitTrafficStatsWithClusterCreateUpdateDelete) {
+  initialize();
+
+  ExpectStats("cluster_1", true, 0, false, 0);
+
+  SendRequestToCluster1AndWaitForResponse();
+  // Cluster_1 trafficStats updated.
+  ExpectStats("cluster_1", true, 1, true, 1);
+
+  UpdateCluster1();
+  test_server_->waitForCounterGe("cluster_manager.cds.update_success", 2);
+
+  // Now the ClusterTrafficStats 'initialized' gauge >= 1, since there is a small race window between
```

Comment:
if this test races we should make the test deterministic using https://github.com/envoyproxy/envoy/blob/main/source/common/common/thread_synchronizer.h

and explicitly cover the race both ways.

Reply:
I think we talked about this before, and we agreed that too much 
The race window is that: 1. in this test, we are making a read on the cluster_1.ClusterTrafficStats.initialized gauge, 
                                           2. in either some worker thread, or the main thread, the threadlocal cluster destructor eventually removed the last reference to the shared_ptr clusterInfoImpl. 

I am not sure if we should latch a checkpoint into lazyinit<T> to capture this, since it's gonna be another unique_ptr cost against the per cluster/route/vhost StatsStruct. 

thoughts?

Reply:
we had an offline conversation. the race window is between testing threads and whatever the last thread that holds the threadlocal clusterInfo object. 
adding a thread_synchronizer is not worthwhile for the testing only purpose, since it has to be added to the per-statsStruct, where the gauge destructs in lazyinit.h.